enhancements/rhcos/rhcos-cloud-marketplace-images.md

```diff
@@ -0,0 +1,289 @@
+
+The `offer` and `publisher` are static values. `Sku` is variable, but deterministic based on the release inputs
+
+##### Azure Stream Representation
+
+An Azure Marketplace extension, `azure-marketplace`, would be added, with unpaid and paid (both North America & EMEA)
+marketplace images listed as child objects: 
+
+```yaml
+"rhel-coreos-extensions": {
+  "azure-disk": {
+    "release": "418.94.202410090804-0",
+    "url": "https://rhcos.blob.core.windows.net/imagebucket/rhcos-418.94.202410090804-0-azure.x86_64.vhd"
+  },
+  "azure-marketplace": {
+    "no-purchase-plan": {
+      "publisher": "RedHat",
+      "offer": "rhcos",
+      "sku": "rhcos",
+      "version": "418.94.202410090804"
+    },
+    "purchase-plan-north-america": {
+      "publisher": "RedHat",
+      "offer": "rh-ocp-worker",
+      "sku": "rh-ocp-worker",
+      "version": "418.94.202410090804"
+    },
+    "purchase-plan-emea": {
+      "publisher": "redhat-limited",
+      "offer": "rh-ocp-worker",
+      "sku": "rh-ocp-worker",
+      "version": "418.94.202410090804"
+    }
+  }
+```
+
+[Hyperv Generation 1 images](https://learn.microsoft.com/en-us/azure/virtual-machines/generation-2#features-and-capabilities)
+would be published, identified by a `gen1` suffix, but would not be included in the RHCOS stream. Any time that the installer
+or users need to select a gen1 image, they would simply add the `gen1` suffix to the SKU identified in the RHCOS stream.  
+
+
+### Risks and Mitigations
+
+TODO
+
+### Drawbacks
+
+TODO
+
+## Open Questions [optional]
+
+`plume cosa2stream` is idempotent. I believe the changes proposed here are sufficient for the idempotency guarantee,
+but the values for marketplace images would change based on the availability of those images. So the same command
+run at two different times would produce different results--but invoking the command multiple times would never
+cause any problems.
+
+## Test Plan
+
+TODO
+
+## Graduation Criteria
+
+TODO
+
+### Dev Preview -> Tech Preview
+
+TODO
+
+### Tech Preview -> GA
+
+TODO
+
+### Removing a deprecated feature
+
+N/A
+
+## Upgrade / Downgrade Strategy
+
+TODO
+
+## Version Skew Strategy
+
+TODO
+
+## Operational Aspects of API Extensions
+
+TODO
+
+## Support Procedures
+
+TODO
+
+## Alternatives
+
+Early feedback on this enhancement proposed splitting marketplace images into a separate file.
+Multiple files solves the issues regarding holding the initial bump PR to wait for marketplace
+images to be uploaded, but there are many assumptions throughout openshift that the stream is contained
+in a single file, such as `openshift-install coreos print-stream-json`, the 
+[coreos configmap manifest](https://github.com/openshift/installer/blob/master/hack/build-coreos-manifest.go),
+& perhaps more. I believe the current approach solves the problems we were trying to solve with this alternative
+but in a way that requires less reworking of existing code and keeps `rhcos.json` as a single canonical file.
```

Comment:
Hmm, even so I still think that having it be a separate file would be cleaner. I may be wrong, but offhand it doesn't seem like it'd require that much work given how simple e.g. the print and configmap code is. One thing that helps is that this is new information, so only things that need to care about Marketplace images need to know how to access it.

Currently, `plume cosa2stream` is purely about transforming some JSON about a build into some other JSON schema. I think it's confusing to have the command now also do cloud queries to find out what the Marketplace images are. It also makes it non-deterministic. But also, having it be separate tools and files managed by separate teams is organizationally just easier.

Reply:
My primary concern with using the separate file is that we would lose having a single canonical source for RHCOS images. We plan on moving all Azure installs to marketplace images, so for example, clients would need to know "I'm looking for Azure images so I don't look in rhcos.json like I do for all other images, I need to look in rhcos-marketplace.json because I know Azure uses marketplace images." 

I'm not sure if there is a way to use json refs or another solution to mitigate this issue (i.e. so that even though the stream is contained in separate json files it can still essentially be deserialized as a single canonical source). I will definitely look into that more (& please lmk if you know). 

I don't think `plume cosa2stream` is the right place--this is my ignorance of cosa. It might be `ore` that would be a more appropriate place for this to live, but I've had some difficulty pinpointing exactly how we publish the current Azure images.

Reply:
Considering this more, I think my concern can be handled by (some sort of) an additional layer of abstraction, like the one I mentioned (json ref) or just a package for loading both streams. I will update the enhancement to use the additional file implementation.

Reply:
>We plan on moving all Azure installs to marketplace images, so for example, clients would need to know "I'm looking for Azure images so I don't look in rhcos.json like I do for all other images, I need to look in rhcos-marketplace.json because I know Azure uses marketplace images."

Users are not supposed to look at the JSON file directly. They should use the `openshift-install coreos print-stream-json` command and we can make that command combine or print the right images in the code.

I agree with Jonathan that this should be a different file.

Reply:
We introduced that command explicitly so that we would not create an interface with the `rhcos.json` file.

Reply:
That WFM. I think keeping it separate in-tree, but together in the cluster is a good balance.